File: PDOO/Material Complementario/Tema5/Series/src/series/Capitulo.java
package series;

import java.util.ArrayList;

/**
 *
 * @author mjfortiz
 */
public class Capitulo {

    private int numero;
    private String titulo;
    private String descripcion;
    private ArrayList<String> elenco;
    private static int contadorCapitulos = 0;

    private Serie miSerie;

    Capitulo() {  //"constructor" sin argumentos, da valores por defecto
        numero = 1;
        titulo = "El Comienzo";
        descripcion = null;
    }

    Capitulo(int numero, String tit, String des) { //constructor que inicializa todos los atributos
        this.numero = numero;
        titulo = tit;
        descripcion = des;
        elenco = new ArrayList<>(); //dar un valor por defecto a algún atributo
    }

    Capitulo(String tit, String des) { //constructor que inicializa atributos
        contadorCapitulos = contadorCapitulos + 1;
        numero = contadorCapitulos;
        titulo = tit;
        descripcion = des;
        elenco = new ArrayList<>(); //dar un valor por defecto a algún atributo
    }

    Capitulo(Capitulo otroCapitulo) { //constructor de copia
        numero = otroCapitulo.numero;
        titulo = otroCapitulo.titulo;
        descripcion = otroCapitulo.descripcion;
        elenco = otroCapitulo.elenco;

    }

    @Override
    public String toString() {
        return "numero de capitulo: " + numero + ", titulo: " + titulo + ", descripcion "
                + descripcion;
    }

    public static void main(String args[]) {
        Capitulo cap = new Capitulo(); // llamando al constructor por defecto, si no hay
        System.out.println(cap);

        Capitulo capA = new Capitulo(); //llamada a constructor sin argumentos
        System.out.println(capA);
        Capitulo capB = new Capitulo();
        System.out.println(capB);

//        Llamando al constructor con tres argumentos
        Capitulo capC = new Capitulo(1, "El caballo", "Se encuentran un caballo");
        System.out.println(capC);
        Capitulo capD = new Capitulo(2, "El lago", "Se montan en una barca");
        System.out.println(capD);

//        Llamando al constructor con dos argumentos que dá valores por defecto
        Capitulo capE = new Capitulo("El bicho", "El bicho le pica");
        System.out.println(capE);
        Capitulo capF = new Capitulo("La medicina", "Se toma la medicina");
        System.out.println(capF);
        Capitulo capG = new Capitulo(8, "La trampa", "Le tienden una trampa");
        System.out.println(capG);

//        Llamando al constructor de copia
        Capitulo capituloCopiado = new Capitulo(capG);
        System.out.println(capituloCopiado);

//        Probando a ver que son dos objetos diferentes, aunque sí compartirán el elenco por ser no invariable
        capG.numero = 99;
        System.out.println(capituloCopiado);
    }

}

File: C1/PDOO/Material Complementario/Tema5/SeriesRuby/lib/capitulo.rb
class Capitulo
  @@contador_numeros=0
#  attr_reader :numero, :titulo, :descripcion
  
#  def initialize  # da valores por defecto sin argumentos
#    @numero=1
#    @titulo="El comienzo"
#    @descripcion= "Llegan al aeropuerto"
#  end
#  
#  def initialize  # si se mantiene este constructor y el anterior, hace caso al último definido antes de usarlo
#    @numero = 4
#    @titulo = nil
#    @descripcion="ksksk"
#  end

#  def initialize (numero, titulo, descripcion) # construtor que da valores de los argumentos
#    @numero=numero
#    @titulo=titulo
#    @descripcion=descripcion
#  end
  

  
#  def initialize (titulo, descripcion) 
#  # constructor con menos argumentos pero da valor por defecto a las variables que faltan
#    @@contador_numeros=@@contador_numeros+1
#    @numero=@@contador_numeros
#    @titulo=titulo
#    @descripcion=descripcion
#    @elenco=Array.new 
#  end
  
#creando un método de clase para que funcione como un constructor, llama al initialize con tres argumentos
#  def self.otroConstructor(numero, descripcion)
#    new(numero, "capitulo interesante", descripcion)
#  end
  
#creando un constructor con valores por defecto en los argumentos
#en este caso, el constructor puede ser usado con dos o con tres argumentos
#  def initialize (numero, titulo, descripcion="siempre lo mismo") 
#    @numero=numero
#    @titulo=titulo
#    @descripcion=descripcion
#  end
#  
# # otra forma de hacer lo mismo
#  def initialize (*args) # cadena de valores
#    @numero, @titulo, @descripcion=*args
#    if args.size<3  # porque en otro caso sería nil
#      @descripcion=' '
#    end
#  end

  #Constructor de copia para usarlo con el initialize de tres argumentos 
  # y con attr_reader de numero, titulo y descripcion
#  def self.constructor_copia(otroCapitulo)
#    new(otroCapitulo.numero, otroCapitulo.titulo, otroCapitulo.descripcion)
#  end
#  
#  def to_s
#    if @titulo==nil
#      @titulo=' '
#    end
#    "numero: "+ @numero.to_s+ ", titulo: " + @titulo + ", descripcion: "+ @descripcion
#  end
  
  
  
# probando si no hay constructor o un constructor sin argumentos
#  capA= Capitulo.new  
#  puts capA

#  probando constructor que da valores a todos los argumentos
#  capB=Capitulo.new(1, "La isla", "llegan a la isla")
#  puts capB

#  probando constructor con menos argumentos 
#  capC=Capitulo.new("La isla", "llegan a la isla")
#  puts capC
#  capD=Capitulo.new("Otra isla", "Se mudan a otra isla")
#  puts capD

# probando constructor en método de clase
#  capE=Capitulo.otroConstructor(6, "Se pone interesante")
#  puts capE
#  capF=Capitulo.new(7, "El bicho", "Le pica el bicho")
#  puts capF

# probando constructor con diferentes argumentos (con = o con *args
#   capJ= Capitulo.new(8, "J", "pasan cosas")
#   puts capJ
#   capK=Capitulo.new(9, "B")  
#   puts capK

#  # probando constructor de copia y constructor con tres argumentos
#  capG= Capitulo.new(10, "G", "se marchan")
#  puts capG
#  capL= Capitulo.constructor_copia(capG)
#  puts capL
  
end
